app/models/token.rb

```diff
@@ -18,32 +18,56 @@ def last_used_date
     end
   end
 
-  # Each test will decrease the number of tokens by one
-  def decrease_tokens
-    if self.remaining > 0
-      self.remaining = self.remaining - 1
-      if self.last_used.nil?
-        self.last_used = DateTime.now
-      end
-      save
-    end
-  end
-
   def reassign_tokens
     assignment = grouping.assignment
-    if DateTime.now < assignment.token_start_date || !grouping.is_valid?
+    if Time.zone.now < assignment.token_start_date || !grouping.is_valid?
       self.remaining = 0
     elsif assignment.unlimited_tokens
-      # grouping has  1 token that is never consumed
+      # grouping has 1 token that is never consumed
       self.remaining = 1
-    elsif last_used.nil? ||
-          (last_used.to_time.to_i + assignment.token_period * 60 * 60 <=
-            DateTime.now.to_time.to_i)
+    elsif self.last_used.nil? || (self.last_used + assignment.token_period.hours) < Time.zone.now
```

Comment:
Yes, I kept the existing semantics, which basically is "X tests every token_period from your last test", I agree it's confusing.
I still like the idea of assigning multiple tokens at once, especially from a group perspective, or simply because you want to allow students to run X tests every day.
Let me think if there's a way to make it work in a nicer way..